homeassistant/components/mqtt/discovery.py

```diff
@@ -206,38 +206,41 @@
                 if value[-1] == TOPIC_BASE and key.endswith('_topic'):
                     payload[key] = "{}{}".format(value[:-1], base)
 
-        # If present, the node_id will be included in the discovered object id
-        discovery_id = '_'.join((node_id, object_id)) if node_id else object_id
-
-        if ALREADY_DISCOVERED not in hass.data:
-            hass.data[ALREADY_DISCOVERED] = {}
-
+        # If present, unique_id is used as the discovered object id. Otherwise,
+        # if present, the node_id will be included in the discovered object id
+        discovery_id = payload.get(
+            'unique_id', '_'.join(
+                (node_id, object_id)) if node_id else object_id)
```

Comment:
One edge case: If a unique ID is *added* to an entity which previously didn't have one, this will create a new entity. It won't happen in practice though, so it's ok and makes the code a lot more readable :P

Also, let's use a space as the `.join` character, as `_` is a valid character for the object id.

Reply:
Fixed.